HW3

Danil-Zaripov opened the pull request:
Completed 3rd homework assignment:
Implemented sorting algorithms for linked lists
Added property based testing using FsCheck.xUnit

gsvgit commented on src/Lib1/Library.fs:
```diff
@@ -1,2 +1,44 @@
+                    x :: (_merge (xs, y :: ys))
+                else
+                    y :: (_merge (x :: xs, ys))
+            | ([], y :: ys) -> y :: ys
```
1. Зачем тут декомпозиция.
2. Кажется, можно в один шаблон свести

gsvgit commented on src/Lib1/Library.fs:
```diff
@@ -1,2 +1,44 @@
+        | x :: xs ->
+            let smaller = List.filter ((>) x) xs
+            let bigger = List.filter ((<=) x) xs
+            (quickSort smaller) @ [ x ] @ (quickSort bigger)
```
Зачем список из одного элемента?

gsvgit commented on test/Lib1Tests/PropertyTests.fs:
```diff
@@ -0,0 +1,19 @@
+module SortTests =
+
+    [<Property>]
+    let CheckBubbleSortAlgorithm (lst: int list) =
```
Надо бы явно зафиксировать крайние случаи.

Danil-Zaripov commented:
Added unit tests.
Improved quick sort and merge sort

gsvgit reviewed (commented):
Ок. А теперь всё то же самое, но на собственноручно написанном типе для списков.

Danil-Zaripov commented:
Implemented my own MyList type. Implemented all the needed functions for the sorting algorithms. Rewritten sorting algorithms for MyList. Tests use MyList.bind for converting from List to MyList and backwards.

gsvgit commented on src/Lib1/Library.fs:
```diff
@@ -1,2 +1,128 @@
+        | Empty -> true
+        | _ -> false
+
+let inline (=.=) hd tl = Cons(hd, tl)
```
Зачем `inline`?

Danil-Zaripov replied:
Больше не `inline`. Если бы это был макрос как в C/C++, то он бы делал то что я хотел, а не был рекомендацией для JIT.

gsvgit commented on src/Lib1/Library.fs:
```diff
@@ -1,2 +1,128 @@
+    let fromFSList xs =
+        xs |> (List.fold (fun st x -> x =.= st) Empty) |> rev
+
+    let bind alg xs = xs |> fromFSList |> alg |> toFSList
```
Кто такой `alg`?

gsvgit commented on src/Lib1/Library.fs:
```diff
@@ -1,2 +1,128 @@
+
+        rev (_append (rev list1) list2)
+
+let inline (@.@) xs ys = MyList.append xs ys
```
Зачем `inline`?

Danil-Zaripov replied:
Больше не `inline`.
